crypto/multisig/src/v2.rs

```diff
@@ -0,0 +1,1259 @@
+        self.threshold
+    }
+
+    /// Verify a message against a multi-signature, returning the list of
+    /// signers that signed it.
+    ///
+    /// NOTE: If a signer appears in multiple groups, it will only be matched
+    /// once, and for the first group it appears in that passed the signing
+    /// threshold (and the ordering of the groups and signatures depends on
+    /// their byte representation since we always sort them). Here is an
+    /// example that demonstrates the implications of this: Assume a signer set,
+    /// that requires 1-of-2 and is sorted, and where each member is a
+    /// m-of-n group where the first requires 2/3 and the second requires
+    /// 3/3:
+    /// 1) A, B, C (threshold 2)
+    /// 2) A, D, E (threshold 3)
+    /// If we are passed signatures A, B, C, D, E we could in theory satisfy the
+    /// threshold of both groups but since A gets "consumed" when matching the
+    /// first group, leaving the second group with only D and E, it will not
+    /// be considered a match. We say we could match all 5 signatures
+    /// in theory, because if we ignored A for the first set, it could still be
+    /// used to match the 2nd set. Such algorithm would require a more
+    /// complex (and slower) implementation, and is not currently supported.
+    pub fn verify<
+        SIG: Clone
+            + Default
+            + Digestible
+            + Eq
+            + Hash
+            + Message
+            + Ord
+            + PartialEq
+            + PartialOrd
+            + Serialize
+            + Signature,
+    >(
+        &self,
+        message: &[u8],
+        multi_sig: &MultiSig<SIG>,
+    ) -> Result<Vec<S>, SignatureError>
+    where
+        S: Verifier<SIG>,
+    {
+        if multi_sig.signatures().len() > MAX_SIGNATURES {
+            return Err(SignatureError::new());
+        }
+
+        // Sort and dedup the list of signatures..
+        let mut signatures = multi_sig.signatures.clone();
+        signatures.sort_by(|a, b| a.as_ref().cmp(b.as_ref()));
+        signatures.dedup();
+
+        // Verify signatures.
+        self.verify_helper(message, &signatures, &[])
+    }
+
+    fn verify_helper<
+        SIG: Clone
+            + Default
+            + Digestible
+            + Eq
+            + Hash
+            + Message
+            + Ord
+            + PartialEq
+            + PartialOrd
+            + Serialize
+            + Signature,
+    >(
+        &self,
+        message: &[u8],
+        signatures: &[SIG],
+        seen_signers: &[S],
+    ) -> Result<Vec<S>, SignatureError>
+    where
+        S: Verifier<SIG>,
+    {
+        // Sort and dedup the list of signers.
+        // While the verification code below should be immune to duplicate signers, the
+        // overhead of deduping them is negligible and being extra-safe is a good idea.
+        let mut signers = self.signers.clone();
+        signers.sort();
+        signers.dedup();
```

Comment:
Continuing this line of reasoning, I think it serves no useful security purpose to deduplicate the list of signers. It is potentially useful to the signers and there is no good reason to prevent it, so its confusing to do this.

If we actually think this should be disallowed, the appropriate place would be to prevent MintConfigTx from being accepted that have duplicate signers, rather than causing the signature verification code to fail in a subtle way without a clear error message explaining the issue.